### DISCUSSION TITLE
lamp-cloud-pro-datasource-colume不能以colume的模式进行业务开发报错

### DISCUSSION THREAD

**Discussion by hellon** [zuihou/zuihou-admin-cloud#255] [02.提问]
描述：
在lamp_base_1中新建表`base_diehl_notice`,表中添加租户ID字段`tenant_id`用于区分不同的租户的数据，生成代码，配置租户，设置权限等后，使用拥有该业务的租户查询表数据时，访问的库是lamp_base_XXXXXXXX, 按说应该访问lamp_base_1的库。
1. `base_diehl_notice`表结构如下：
```
CREATE TABLE `base_diehl_notice` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT COMMENT '公告ID',
  `title` varchar(50) COLLATE utf8mb4_unicode_ci NOT NULL COMMENT '公告标题',
  `content` text COLLATE utf8mb4_unicode_ci NOT NULL COMMENT '公告内容',
  `type` tinyint(4) NOT NULL COMMENT '公告类型（1通知 2公告）',
  `status` tinyint(4) NOT NULL DEFAULT '0' COMMENT '公告状态（0正常 1关闭）',
  `created_by` bigint(20) NOT NULL DEFAULT '0' COMMENT '创建者',
  `create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT '创建时间',
  `updater` varchar(64) COLLATE utf8mb4_unicode_ci DEFAULT '' COMMENT '更新者',
  `update_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT '更新时间',
  `deleted` bit(1) NOT NULL DEFAULT b'0' COMMENT '是否删除',
  `tenant_id` bigint(20) NOT NULL DEFAULT '0' COMMENT '租户编号',
  PRIMARY KEY (`id`) USING BTREE
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci COMMENT='通知公告表';
```

2. 代码生成配置如下：
![code](https://user-images.githubusercontent.com/15033782/227827246-601892c5-60c3-4860-9677-20e1383df79a.png)

问题：
1. 这次的业务是多个租户公用一个数据库，然后以colume的模式进行数据区分，租户ID该如何设置
2. 购买给的文档中没有针对如何开发这种以colume模式的文档，是否有这样的文档，在哪里查看

**Comment by zuihou** [comment]
1. datasource-column 模式，用 datasource隔离大租户、column隔离小租户（门店）。
    下图中，红色框表示 大租户，租户1、租户2 分别使用 lamp_base_1、lamp_base_2。 蓝色框是 lamp_base_1/2 库中的 小租户， 用tenant_id 字段隔离。

    开发步骤：
     1. 在service层加@DS注解，用以切换 lamp_base_1 或 lamp_base_2 。 其中@DS注解选择base_1还是base_2，取决于`ContextUtil.getTenantBasePoolName()`
     2. 在mapper层加 @TenantLine 注解， 用以在sql中拼接 tenant_id = xxx， xxx参数取决于 `ContextUtil.getCurrentCompanyId()`




     ![大租户小门店](https://user-images.githubusercontent.com/8500410/227834345-0bbcdf61-849b-4504-9e20-aab7526d4cad.png)


3. 没有

**Comment by hellon** [comment]
以小租户的模式开发：在mapper层加 @TenantLine 注解， 用以在sql中拼接 tenant_id = xxx， xxx参数取决于 ContextUtil.getCurrentCompanyId()
在进行数据添加的时候，这个tenant_id不能自动设置到VO里面吗，如果手动设置的话需要重写系统提供的SaveController.save()方法

**Reply by zuihou** [reply]
在写入数据时，确保ContextUtil.getCurrentCompanyId()有值，save时 就能自动赋值

**Comment by hellon** [comment]
我使用column的方式添加，进行debug，发现拼接使用的是created_org_id,且系统中其他默认的都使用的是created_org_id。
这个如何进行调整修改：
![tenantIdColumn](https://user-images.githubusercontent.com/15033782/230830577-e4b247b9-8089-4a3f-9efe-1f4bb7db7794.png)
查询的时候也是使用的created_org_id:
![image](https://user-images.githubusercontent.com/15033782/230830865-83f64bbc-abbf-43e6-8258-f2359cdf4ceb.png)

**Reply by zuihou** [reply]
修改配置文件，使databaseProperties.getTenantIdColumn参数变成你想要的字段。
